src/main/java/com/puppycrawl/tools/checkstyle/site/ViolationMessagesMacro.java

```diff
@@ -0,0 +1,274 @@
+        }
+        catch (CheckstyleException ex) {
+            throw new MacroExecutionException("Couldn't load checkstyle modules", ex);
+        }
+    }
+
+    /**
+     * Returns the checkstyle messages for the given class.
+     *
+     * @param clss the class.
+     * @return the checkstyle messages.
+     * @throws MacroExecutionException if the messages could not be retrieved.
+     */
+    private static Set<Field> getCheckMessageKeyFields(Class<?> clss)
+            throws MacroExecutionException {
+        final Set<Field> checkstyleMessages = new HashSet<>();
+
+        final Field[] fields = clss.getDeclaredFields();
+
+        for (Field field : fields) {
+            if (field.getName().startsWith("MSG_")) {
+                checkstyleMessages.add(field);
+            }
+        }
+
+        // deep scan class through hierarchy
+        final Class<?> superModule = clss.getSuperclass();
+
+        if (superModule != null && shouldScanDeepClassForMessages(superModule)) {
+            checkstyleMessages.addAll(getCheckMessageKeyFields(superModule));
+        }
+
+        // special cases that require additional classes
+        try {
+            if (clss == RegexpMultilineCheck.class) {
+                checkstyleMessages.addAll(getCheckMessageKeyFields(Class
+                    .forName("com.puppycrawl.tools.checkstyle.checks.regexp.MultilineDetector")));
+            }
+            else if (clss == RegexpSinglelineCheck.class
+                    || clss == RegexpSinglelineJavaCheck.class) {
+                checkstyleMessages.addAll(getCheckMessageKeyFields(Class
+                    .forName("com.puppycrawl.tools.checkstyle.checks.regexp.SinglelineDetector")));
+            }
+        }
+        catch (ClassNotFoundException ex) {
+            throw new MacroExecutionException("Couldn't find class", ex);
+        }
+
+        return checkstyleMessages;
+    }
+
+    /**
+     * Should the class be deep scanned for messages.
+     *
+     * @param superModule The class to examine.
+     * @return {@code true} if the class should be deeply scanned.
+     */
+    private static boolean shouldScanDeepClassForMessages(Class<?> superModule) {
+        return superModule == AbstractNameCheck.class
+                || superModule == AbstractAccessControlNameCheck.class
+                || superModule == AbstractParenPadCheck.class
+                || superModule == AbstractSuperCheck.class;
+    }
+
+    /**
+     * Returns the message keys for the given fields.
+     *
+     * @param messageKeyFields the fields.
+     * @param instance the instance of the module.
+     * @return the message keys.
+     * @throws MacroExecutionException if the message keys could not be retrieved.
+     */
+    private static Set<String> getMessageKeys(Set<Field> messageKeyFields, Object instance)
+            throws MacroExecutionException {
+        final Set<String> messageKeys = new TreeSet<>();
```

Comment:
Why TreeSet?

Reply:
We want to preserve the order of the keys. Using HashSet messed up the order and failed XdocsJavadocsTest. That's what's being done in `XdocsPagesTest`

https://github.com/checkstyle/checkstyle/blob/0c2ab5338db569b4699d5d20121d1bd508394eb9/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1381-L1389

Reply:
Hm, why are we forced to use `Set` at all?

Reply:
I got rid of the `Set` and just used a `List`.

@rnveach could you please tell if using a `Set` here is a must? Could there be some duplicates or?

Reply:
No duplicates are allowed in messages

Reply:
> No duplicates are allowed in messages

Please elaborate on this; do we have some way to enforce this or do we make sure by Set API?

Reply:
Set, will hide duplicates to make illusion in html.
Better to use List to simply print as is.

The only extra benefit from fancier collection I see, is ordering.
It will be good to order alphabetically in html.
But we can do this later on, after transition to templates is done.

Reply:
> It will be good to order alphabetically in html.

It is already done in `ViolationMessagesMacro::createListOfMessages`
```java
        Collections.sort(messageKeys);
```

Reply:
You can't use an object that is based on hashcode. It is JVM specific and different runs can generate different orders. We can't use insertion order here because reflection is not guaranteed an order and we want to ensure our documentation is consistent. This is why we originally used `TreeSet`.

I recommend keeping this a copy/paste and just create new issues to fix anything.

Reply:
@rnveach awesome explanation, thanks a lot. @stoyanK7 please restore copy pasta and add a comment about why we use treeset

Reply:
@nrmancuso Done at https://github.com/checkstyle/checkstyle/pull/13505.